packages/block-library/src/query/edit/query-content.js

```diff
@@ -79,14 +78,18 @@ export default function QueryContent( {
 			updateQuery( newQuery );
 		}
 	}, [ query.perPage, postsPerPage, inherit ] );
-	// We need this for multi-query block pagination.
-	// Query parameters for each block are scoped to their ID.
+
+	// Each instance of the block needs to have a unique queryId so that pagination works
+	// when multiple instances of this block are present. `instanceId` is the index of
+	// the block on the page and is stable across saves with automatic handling for
+	// re-indexing when these blocks are added, removed, or moved. Depending on
+	// what it's set to it might be changed during block editing but that won't
+	// negatively impact the functionality and pagination will still work.
 	useEffect( () => {
-		if ( ! Number.isFinite( queryId ) ) {
-			__unstableMarkNextChangeAsNotPersistent();
-			setAttributes( { queryId: instanceId } );
-		}
-	}, [ queryId, instanceId ] );
+		__unstableMarkNextChangeAsNotPersistent();
+		setAttributes( { queryId: instanceId } );
+	}, [] );
```

Comment:
Do we not want to add `instanceId` as a dependency to this `useEffect` hook?

Reply:
Initially I didn't @tjcafferkey because this value is persistent during component re-mounting. It seemed reasonable to just declare that the hook should only run on mounting. Adding the `instanceId` would make the linter happier though so I've gone ahead and added it.